File: week1/Programming-Assignment-1/check_brackets_in_code/check_brackets.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

struct Bracket {
    Bracket(char type, int position):
    type(type),
    position(position)
    {}

    bool Matchc(char c) {
        if (type == '[' && c == ']')
            return true;
        if (type == '{' && c == '}')
            return true;
        if (type == '(' && c == ')')
            return true;
        return false;
    }

    char type;
    int position;
};

int main() {
    std::string text;
    getline(std::cin, text);

    std::stack <Bracket> opening_brackets_stack;
    for (int position = 0; position < text.length(); ++position) {
        char next = text[position];
        if (next == '(' || next == '[' || next == '{') {
            // Process opening bracket
            opening_brackets_stack.push(Bracket(next, position + 1));
        }

        if (next == ')' || next == ']' || next == '}') {
            // Process closing bracket
            if (opening_brackets_stack.empty()) {
                opening_brackets_stack.push(Bracket(next, position + 1));
                break;
            }
            Bracket bracket = opening_brackets_stack.top();
            if (bracket.Matchc(next)) {
                opening_brackets_stack.pop();
            } else {
                opening_brackets_stack.push(Bracket(next, position + 1));
                break;
            }
        }
    }

    // Printing answer, write your code here
    if (opening_brackets_stack.empty()) {
        cout << "Success" << endl;
    } else  {
        cout << opening_brackets_stack.top().position << endl;
    }

    return 0;
}

File: week1/Programming-Assignment-1/check_brackets_in_code/check_brackets.py
# python3

from collections import namedtuple

Bracket = namedtuple("Bracket", ["char", "position"])


def are_matching(left, right):
    return (left + right) in ["()", "[]", "{}"]


def find_mismatch(text):
    opening_brackets_stack = []
    for i, char in enumerate(text):
        if char in "([{":
            opening_brackets_stack.append(Bracket(char, i+1))

        if char in ")]}":
            if len(opening_brackets_stack) == 0:
                opening_brackets_stack.append(Bracket(char, i+1))
                break
            b = opening_brackets_stack[-1]
            if are_matching(b.char, char):
                opening_brackets_stack.pop()
            else:
                opening_brackets_stack.append(Bracket(char, i+1))
                break
    if len(opening_brackets_stack) == 0:
        return "Success"
    else:
        return opening_brackets_stack[-1].position


def main():
    text = input()
    mismatch = find_mismatch(text)
    # Printing answer, write your code here
    print(mismatch)


if __name__ == "__main__":
    main()
